client/tinychain/hosted_ml/nn.py

```diff
@@ -198,3 +130,55 @@ def create(cls, schema):
 
         layers = Tuple([DNNLayer.create(*layer_schema) for layer_schema in schema])
         return cls(layers)
+
+
+#TODO: Convalution for ConvLayer.operator()
+class Convolution(Operator):
+    def __init__(self, params, inputs):
+        Operator.__init__(self, params.weights, params.bias)
+        self.params = params
+        self.inputs = inputs
+
+    def forward(self):
+        # c_i, h_i, w_i = self.params._inputs_shape
+        # out_c, h_f, w_f = self.params._filter_shape
+        self.b_i = self.inputs.shape[0]
+        self.h_out = int((self.params.h_i - self.params.h_f + 2 * self.params._padding) / self.params._stride + 1)
+        self.w_out = int((self.params.w_i - self.params.w_f + 2 * self.params._padding) / self.params._stride + 1)
+
+        assert self.h_out
+        assert self.w_out
+
+        pad_matrix = Dense.zeros([self.b_i, self.params.c_i, self.params.h_i + self.params._padding * 2, self.params.w_i + self.params._padding * 2])
+        im2col_matrix = []
+        for i in range(self.h_out):
+            for j in range(self.w_out):
+                im2col_matrix.append(pad_matrix[:, :, i:i + self.params.h_f, j:j + self.params.w_f].reshape([self.params.c_i * self.params.h_f * self.params.w_f, self.b_i]))
+        im2col_concat = Tensor(After(pad_matrix[:, :, self.params._padding:(self.params._padding + self.params.h_i), self.params._padding:(self.params._padding + self.params.w_i)].write(self.inputs.copy()), Dense.concatenate(im2col_matrix, 0)))
+        self.im2col_matrix = Tensor(After(im2col_concat, im2col_concat.reshape([self.b_i * self.params.h_out * self.params.w_out, self.params.c_i * self.params.h_f * self.params.w_f]).transpose()))
+        w_col = self.subject.reshape([self.params.out_c, self.params.c_i * self.params.h_f * self.params.w_f])
+        in2col_multiply = (w_col @ self.im2col_matrix + self.arg).reshape([self.params.out_c, self.params.h_out, self.params.w_out, self.b_i])
+        output = Tensor(in2col_multiply.copy().transpose([3, 0, 1, 2]))
+
+        if self._activation:
+            return self._activation(output)
+
+        return output
+
+    def backward(self, dl):
```

Comment:
There should be a way to do this without depending on the value of `dl` in `backward`. I added some ideas here: https://github.com/haydnv/neural_net.py/pull/3

Feel free to implement the convolution operator in the prototype version before continuing work here.